swarm/network/kademlia.go

```diff
@@ -627,68 +634,42 @@ func NewPeerPotMap(kadMinProxSize int, addrs [][]byte) map[string]*PeerPot {
 	ppmap := make(map[string]*PeerPot)
 
 	for i, a := range addrs {
-
 		// set to proxbin depth when all nn-peers are found
-		pl := 256
-
+		depth := depthForPot(np, kadMinProxSize, a)
 		// upon entering a new iteration
 		// this will hold the value the po should be
 		// if it's one higher than the po in the last iteration
-		prev := 256
+		prevPo := 256
 
 		// all bins outside proxbin depth with no peers
 		var emptyBins []int
 
 		// all nn-peers
 		var nns [][]byte
 
-		// used to skip empty bins immediately after nn-peers
-		depthTraversed := false
-
-		np.EachNeighbour(addrs[i], pof, func(val pot.Val, po int) bool {
-			a := val.([]byte)
-
-			// 256 is self. We are selfless
-			if po == 256 {
+		np.EachNeighbour(a, pof, func(val pot.Val, po int) bool {
+			addr := val.([]byte)
+			if bytes.Equal(pot.ToBytes(val), a) {
 				return true
 			}
 
-			// if first nn-peer or peer in same bin as last
-			if pl == 256 || pl == po {
-				nns = append(nns, a)
-			}
-
-			// if true then all nn-bins have been filled
-			// start counting pl and set prev to current po initially (which will skip next block)
-			if pl == 256 && len(nns) >= kadMinProxSize {
-				pl = po
-				prev = po
+			// iterate through the neighbours, going from the closest to the farthest
+			// we calculate the nearest neighbours that should be in the set
+			// depth in all bins that are higher or equal than depth there are
+			// at least minproxbin size connected and depth-1 is not empty
+			if po >= depth {
+				nns = append(nns, addr)
+				prevPo = depth - 1
```

Comment:
Why is this needed in addition to line 668? Can the logic be merged?

Reply:
you have to return here otherwise you'd get empty bins inside the neighbourhood depth. i don't see how the two could be merged. seems about right.